text/3305-static-in-pattern.md

```diff
@@ -0,0 +1,285 @@
+        ENOENT => make_it_work(),
+        _ => do_different_stuff(),
+    }
+}
+```
+
+Because Rust patterns don't support statics, all the `match` expressions in the standard library that refer to POSIX constants would currently need to be rewritten to accommodate Cosmopolitan.
+
+```rust
+// stdlib code adapted for cosmopolitan
+
+use libc::*;
+
+fn process_error(error_code: i32) {
+    if error_code == EINVAL {
+        do_stuff();
+    } else if error_code == ENOSYS {
+        panic!("oh noes");
+    } else if error_code == ENOENT {
+        make_it_work();
+    } else {
+        do_different_stuff();
+    }
+}
+```
+
+Needless to say, this is unlikely to ever be upstreamed. Allowing statics in patterns would solve this use-case much more cleanly.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Rust patterns can refer to constants.
+
+```rust
+const EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+With this feature, they can refer to statics as well.
+
+```rust
+static EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+Mutable statics are not allowed, however. Patterns can't reference information that can change at runtime, and also can't be `unsafe`.
+
+```rust
+
+static mut EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        // ERROR can't refer to mutable statics in patterns
+        /* EVERYTHING => println!("have all of it"), */
+        _ => println!("need moar"),
+    }
+}
+```
+
+Statics from `extern` blocks are allowed, but they must be marked as trusted using the (not-yet-implemented) [trusted external statics](https://github.com/rust-lang/lang-team/issues/149) feature.
+
+```rust
+extern "C" {
+    #[unsafe(trusted_extern)]
+    static EVERYTHING: i32;
+}
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+For a static to be eligible for use in a pattern, it must:
+
+- not be marked `mut`
+- not be marked `#[thread_local]`
+- not come from an extern block, unless it is marked as safe to use with the [trusted external statics](https://github.com/rust-lang/lang-team/issues/149) feature
+- have a type that satisfies the structural match rules, as described in [RFC 1445](1445-restrict-constants-in-patterns.md), but without any allowances for backward compatibility like there are for consts (e.g., floating point numbers in patterns) . These rules exclude all statics with interior mutability.
+
+Static patterns match exactly when a const pattern with a const of the same type and value would match.
```

Comment:
This is a very subtle guarantee, since the way these patterns are treated will have to be completely different. Matching `const`s relies heavily on knowing their value. `const` will be taken apart to a 'valtree' that is matched on field-by-field, as least if the type permits, whereas `extern static` will necessarily be compared via `PartialEq`. This may or may not produce the same result, depending on how `PartialEq` is implemented.

Above you reference the structural match RFC, but the rules in that RFC have little to do with what the compiler actually does. The real rules are quite the mess and need to be cleaned up, see https://github.com/rust-lang/rust/issues/74446. But it seems inevitable that at least in unsafe code it will be possible to have "structural match" types that actually violate the structural match contract, and we need to say what happens then. This is not an operational requirement, so it probably cannot become UB.

Reply:
If future relaxations of the structural match requirements are somehow impossible to port to static matches, we can always require the stricter rules for static patterns while loosening them everywhere else. It's impossible to say more without knowing what such relaxations might look like.

Reply:
"Future relaxation" makes it sound like we currently have some rules that we understand, and we want to change them. That is not the case. As I said, what the compiler does and what the RFC says are already different, and last time this came up, there was little interest in adjusting the compiler to the spec -- rather, the spec will be adjusted to the compiler.

So, you simply cannot use that RFC as a basis for further language evolution.

Reply:
When I referred to the structural match RFC, I was mostly concerned about the guarantee that, for some local `x` and `const` `Y`, `match x { Y => true, _ => false }` evaluates to `true` iff `PartialEq::eq(&x, &Y)` evaluates to `true`. Do you know of a counter-example to this that works in stable Rust?

Reply:
We have patterns where `PartialEq::eq(&x, &Y)` doesn't even compile, see e.g. [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=486b60ce80768be393d154eaf6a7c2ef) from the issue that I linked twice already (https://github.com/rust-lang/rust/issues/74446).

Reply:
> We have patterns where `PartialEq::eq(&x, &Y)` doesn't even compile

OK, I definitely need to account for that then. But are there cases where it does compile, but gives a different answer?

(Also, is "comparing function pointers is wierd" just a potentially-fixable LLVM quirk, or is there some deep reason behind it?)

Reply:
> OK, I definitely need to account for that then. But are there cases where it does compile, but gives a different answer?

No, I don't think so. But I already explained why that is not sufficient grounds to claim that static and const matching are equivalent.

> (Also, is "comparing function pointers is wierd" just a potentially-fixable LLVM quirk, or is there some deep reason behind it?)

It's because of `unnamed_addr`, see e.g. https://github.com/rust-lang/rust/issues/70861